x/ccv/provider/keeper/msg_server.go

```diff
@@ -174,3 +174,62 @@ func (k msgServer) SubmitConsumerDoubleVoting(goCtx context.Context, msg *types.
 
 	return &types.MsgSubmitConsumerDoubleVotingResponse{}, nil
 }
+
+func (k msgServer) OptIn(goCtx context.Context, msg *types.MsgOptIn) (*types.MsgOptInResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	valAddress, err := sdk.ConsAddressFromBech32(msg.ProviderAddr)
+	if err != nil {
+		return nil, err
+	}
+	providerAddr := types.NewProviderConsAddress(valAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	if msg.ConsumerKey != "" {
+		k.Keeper.HandleOptIn(ctx, msg.ChainId, providerAddr, &msg.ConsumerKey)
+		ctx.EventManager().EmitEvents(sdk.Events{
+			sdk.NewEvent(
+				ccvtypes.EventTypeOptIn,
+				sdk.NewAttribute(types.AttributeProviderValidatorAddress, msg.ProviderAddr),
```

Comment:
is there a reason to prefer having an opt in event with both fields? could also emit an optIn and an 	AssignConsumerKey event instead

Reply:
Agree. You could also have the `types.AttributeConsumerConsensusPubKey` of the event be an empty `string`, if a consumer key isn't passed.

Reply:
> could also emit an optIn and an AssignConsumerKey event instead
Emitting an `AssignConsumerKey` event as well might be problematic because a relayer might consider that a `MsgAssignConsumerKey` message was sent independently which is not the case. 

Followed @sainoe approach on passing an empty `string`.